drivers/adc/adc_tla2021.c

```diff
@@ -0,0 +1,330 @@
+#define REG_CONFIG_MUX_pos  12 /* TLA2024 Only */
+#define REG_CONFIG_OS_pos   15
+#define REG_CONFIG_OS_msk   (BIT_MASK(1) << REG_CONFIG_OS_pos)
+
+#define ACQ_THREAD_PRIORITY   CONFIG_ADC_TLA2021_ACQUISITION_THREAD_PRIORITY
+#define ACQ_THREAD_STACK_SIZE CONFIG_ADC_TLA2021_ACQUISITION_THREAD_STACK_SIZE
+
+typedef int16_t tla2021_reg_data_t;
+typedef uint16_t tla2021_reg_config_t;
+
+struct tla2021_config {
+	const struct i2c_dt_spec bus;
+	k_tid_t acq_thread_id;
+};
+
+struct tla2021_data {
+	const struct device *dev;
+	struct adc_context ctx;
+	struct k_sem acq_lock;
+	tla2021_reg_data_t *buffer;
+	tla2021_reg_data_t *repeat_buffer;
+
+	/*
+	 * Shadow register
+	 */
+	tla2021_reg_config_t reg_config;
+};
+
+static int read_(const struct device *dev, uint8_t reg, uint16_t *value)
+{
+	int ret;
+
+	const struct tla2021_config *config = dev->config;
+	uint8_t value_[2];
+
+	ret = i2c_write_read_dt(&config->bus, &reg, sizeof(reg), value_, sizeof(value_));
+	if (ret) {
+		return ret;
+	}
+
+	*value = sys_get_be16(value_);
+
+	return 0;
+}
+
+static int write_(const struct device *dev, uint8_t reg, uint16_t value)
+{
+	int ret;
+
+	const struct tla2021_config *config = dev->config;
+	uint8_t value_[3] = {reg};
+
+	sys_put_be16(value, &value_[1]);
+
+	ret = i2c_write_dt(&config->bus, value_, sizeof(value_));
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline int read_data(const struct device *dev, tla2021_reg_data_t *value)
+{
+	return read_(dev, REG_DATA, value);
+}
+
+static inline int read_config(const struct device *dev, tla2021_reg_config_t *value)
+{
+	return read_(dev, REG_CONFIG, value);
+}
+
+static inline int write_config(const struct device *dev, tla2021_reg_config_t value)
+{
+	return write_(dev, REG_CONFIG, value);
+}
+
+static int tla2021_channel_setup(const struct device *dev, const struct adc_channel_cfg *cfg)
+{
+	if (cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid gain");
+		return -EINVAL;
+	}
+
+	if (cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Invalid reference");
+		return -EINVAL;
+	}
+
+	if (cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Invalid acquisition time");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int tla2021_start_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	if (seq->channels != BIT(0)) {
+		LOG_ERR("Selected channel(s) not supported: %x", seq->channels);
+		return -EINVAL;
+	}
+
+	if (seq->resolution != 12) {
+		LOG_ERR("Selected resolution not supported: %d", seq->resolution);
+		return -EINVAL;
+	}
+
+	if (seq->oversampling) {
+		LOG_ERR("Oversampling is not supported");
+		return -EINVAL;
+	}
+
+	if (seq->calibrate) {
+		LOG_ERR("Calibration is not supported");
+		return -EINVAL;
+	}
+
+	if (!seq->buffer) {
+		LOG_ERR("Buffer invalid");
+		return -EINVAL;
+	}
+
+	const size_t num_extra_samples = seq->options ? seq->options->extra_samplings : 0;
+	const size_t num_samples = (1 + num_extra_samples) * POPCOUNT(seq->channels);
```

Comment:
These are normally at the top of the function unless there's a good reason to declare them mid way, I'd suggest moving them. Same for `*data` below.

Reply:
Don't really care ;-). Therefore: Done.